### ISSUE TITLE
dyld: lazy symbol binding failed: Symbol not found: _uv_dlopen when trying to load a native module on an M1 mac

### ISSUE THREAD

**Issue by bergundy** [laverdet/isolated-vm#238]
We're using NativeModule in order to expose v8 promise hooks to the isolate.
It's been working fine on all platforms until now except for mac with an M1 processor.
The NativeModule constructor causes the process to exit due to SIGABRT.

Reference:
https://github.com/temporalio/sdk-node/blob/e6c282628956f336e3ac47e5b5f6514c48c847f1/packages/worker/src/workflow.ts#L38

**Comment by bergundy** [comment]
Looks like the symbol is indeed undefined:
`nm node_modules/isolated-vm/out/isolated_vm.node | grep _uv_dlopen`

```
                 U _uv_dlopen
```

**Comment by bergundy** [comment]
I seemed to have made it work by adding:
```
'libraries': ['-L /opt/homebrew/lib', '-luv'],
```
to the isolated_vm target in the gyp file.
I don't know why we need to explicitly do this for arm.

**Comment by laverdet** [comment]
The output of the following commands would be helpful:
```
otool -L $(which node)
nm -g $(which node)
```

```
# from git checkout, without the -luv workaround you mentioned
node-gyp rebuild -v
nm -g out/isolated_vm.node
```

**Comment by bergundy** [comment]
Sure, thanks for the quick response.
[report.txt](https://github.com/laverdet/isolated-vm/files/6305364/report.txt)

**Comment by laverdet** [comment]
It's really strange because node is still exporting the rest of libuv:
```
nm -g $(which node)
[...]
00000001008b9c70 T _uv_accept
00000001008b03bc T _uv_async_init
00000001008b0514 T _uv_async_send
00000001008b0ac0 T _uv_backend_fd
00000001008b0ac8 T _uv_backend_timeout
00000001008bb6e8 T _uv_barrier_destroy
00000001008bb464 T _uv_barrier_init
```

I diff'd the output of yours vs mine and the only public uv APIs you're missing is 4 uv_dl* function definitions.

I'm not really sure if this is an issue with brew or with nodejs but I'm leaning towards brew. The next step I would take would be to compile and install nodejs from source on the machine using default `./configure` settings and see if the issue is still there. I know package managers can sometimes make really opinionated and frustrating linking decisions.

By the way regarding #201 in my experience this kind of issue ends up being super common when you bring binary releases into the mix :)

**Comment by bergundy** [comment]
I've installed node 14 with nvm and I see that `_uv_dl*` symbols are exported.
```shell
$ nm -g $(which node) | grep _uv_dl
00000001008aae18 T _uv_dlclose
00000001008aaeb8 T _uv_dlerror
00000001008aad98 T _uv_dlopen
00000001008aae50 T _uv_dlsym
```

But something's wrong with this installation, I'm getting process out of memory errors.
I'll see if I can get a proper nodejs installation on my mac.

Thanks!

Also re #201, I know that this is a pain. For the Temporal SDK we decided to take the pain on ourselves instead of letting our users experience it. I totally understand where you're coming from.

**Comment by bergundy** [comment]
It's an issue with the node version from brew.
I installed node 15 with nvm from source and it works.

**Comment by carlocab** [comment]
Apologies for the inconvenience here. The Node from `nvm` statically links `libuv` (along with everything else) and so exports the symbols that you need. Homebrew Node, on the other hand, links with `libuv` dynamically, which is why you need to add the flags you mention here: https://github.com/laverdet/isolated-vm/issues/238#issuecomment-818786255

**Comment by carlocab** [comment]
> it's spooky that the rest of `_uv_*` symbols are exported.. it's just the 4 `_uv_dl_*` symbols which aren't exported.

That is quite strange. I spun up a Docker container and installed Node 16 from NodeSource, and it looks like this build of Node doesn't export any `_uv_*` symbols. (I don't use Linux, though, so I could've just done this wrong.)

> This definitely feels like the wrong place to fix the issue though.

Usually, this is where a project ships `pkg-config` files or a `foo-config`-type executable to simplify linking, but Node doesn't do either of those. I don't really use Node, though, so I can't really comment on why this is the case.

**Comment by laverdet** [comment]
> That is quite strange. I spun up a Docker container and installed Node 16 from NodeSource, and it looks like this build of Node doesn't export any _uv_* symbols. (I don't use Linux, though, so I could've just done this wrong.)

That's not the behavior I'm seeing. On Debian from NodeSource I have all the uv_ symbols. We can verify directly from the source as well via:
```
$ wget https://deb.nodesource.com/node_16.x/pool/main/n/nodejs/nodejs_16.1.0-1nodesource1_amd64.deb
$ ar x nodejs_16.1.0-1nodesource1_amd64.deb
$ tar xf data.tar.xz
$ nm ./usr/bin/node | grep uv_dl
000000000158b9b0 T uv_dlclose
000000000158ba60 T uv_dlerror
000000000158b930 T uv_dlopen
000000000158b9f0 T uv_dlsym
```

**Comment by carlocab** [comment]
You're right -- I had the wrong `grep` query. I had a stray `_`, since these symbols are prefixed with `_` on macOS. Mea culpa. Looking at the install log, though, I suspect these binaries are statically linked, though, so they're not the right basis for comparison.

**Comment by bergundy** [comment]
@laverdet PTAL at my comment in the brew issue: https://github.com/Homebrew/homebrew-core/issues/76893#issuecomment-836612506.

**Comment by bergundy** [comment]
FTR this is resolved in node 16.6 and perhaps 16.5 too but I didn't verify.
Thanks!
